service/root.go

```diff
@@ -0,0 +1,60 @@
+package service
+
+import (
+	"net/http"
+	"time"
+
+	"github.com/go-chi/chi/v5"
+	"github.com/go-chi/chi/v5/middleware"
+	"github.com/riandyrn/otelchi"
+	"go.opentelemetry.io/otel/trace"
+
+	"github.com/zenledger-io/zazen/httpx"
+	"github.com/zenledger-io/zazen/telemetry"
+)
+
+var (
+	defaultRequestTimeout = 60 * time.Second
+)
+
+func newRoot(tp trace.TracerProvider, cfg Config) http.Handler {
+	root := chi.NewRouter()
+
+	timeout := cfg.RequestTimeout
+	if timeout == 0 {
+		timeout = defaultRequestTimeout
+	}
+
+	root.Use(middleware.Timeout(timeout))
+	root.Use(middleware.RealIP)
+	root.Use(middleware.RequestID)
+
+	root.Use(otelchi.Middleware(cfg.Name,
```

Comment:
I see logs come through the span, will the logs all have the same uuid so we can trace them through a request?

Reply:
The logs are a part of a span, which means they are emitted as an event on the span itself.

Run `examples/api` and then `curl localhost:8111/v1/birds` will emit the following span (note how the log message is exported as an event):

```json
{
        "Name": "/v1/birds",
        "SpanContext": {
                "TraceID": "e69e4d9ffdadaf6929969b2940ecdc5e",
                "SpanID": "b821cbc7f8b1c639",
                "TraceFlags": "01",
                "TraceState": "",
                "Remote": false
        },
        "Parent": {
                "TraceID": "00000000000000000000000000000000",
                "SpanID": "0000000000000000",
                "TraceFlags": "00",
                "TraceState": "",
                "Remote": false
        },
        "SpanKind": 2,
        "StartTime": "2022-05-03T19:28:58.196681-06:00",
        "EndTime": "2022-05-03T19:28:58.2975965-06:00",
        "Attributes": [
                {
                        "Key": "net.transport",
                        "Value": {
                                "Type": "STRING",
                                "Value": "ip_tcp"
                        }
                },
                {
                        "Key": "net.peer.ip",
                        "Value": {
                                "Type": "STRING",
                                "Value": "::1"
                        }
                },
                {
                        "Key": "net.peer.port",
                        "Value": {
                                "Type": "INT64",
                                "Value": 61869
                        }
                },
                {
                        "Key": "net.host.name",
                        "Value": {
                                "Type": "STRING",
                                "Value": "localhost"
                        }
                },
                {
                        "Key": "net.host.port",
                        "Value": {
                                "Type": "INT64",
                                "Value": 8111
                        }
                },
                {
                        "Key": "http.method",
                        "Value": {
                                "Type": "STRING",
                                "Value": "GET"
                        }
                },
                {
                        "Key": "http.target",
                        "Value": {
                                "Type": "STRING",
                                "Value": "/v1/birds"
                        }
                },
                {
                        "Key": "http.server_name",
                        "Value": {
                                "Type": "STRING",
                                "Value": "the-birds"
                        }
                },
                {
                        "Key": "http.user_agent",
                        "Value": {
                                "Type": "STRING",
                                "Value": "curl/7.77.0"
                        }
                },
                {
                        "Key": "http.scheme",
                        "Value": {
                                "Type": "STRING",
                                "Value": "http"
                        }
                },
                {
                        "Key": "http.host",
                        "Value": {
                                "Type": "STRING",
                                "Value": "localhost:8111"
                        }
                },
                {
                        "Key": "http.flavor",
                        "Value": {
                                "Type": "STRING",
                                "Value": "1.1"
                        }
                },
                {
                        "Key": "http.route",
                        "Value": {
                                "Type": "STRING",
                                "Value": "/v1/birds"
                        }
                },
                {
                        "Key": "http.status_code",
                        "Value": {
                                "Type": "INT64",
                                "Value": 200
                        }
                }
        ],
        "Events": [
                {
                        "Name": "log",
                        "Attributes": [
                                {
                                        "Key": "log.severity",
                                        "Value": {
                                                "Type": "STRING",
                                                "Value": "DEBUG"
                                        }
                                },
                                {
                                        "Key": "log.message",
                                        "Value": {
                                                "Type": "STRING",
                                                "Value": "inside birds handler"
                                        }
                                },
                                {
                                        "Key": "code.function",
                                        "Value": {
                                                "Type": "STRING",
                                                "Value": "github.com/zenledger-io/zazen/telemetry.(*otelZapLog).Debug"
                                        }
                                },
                                {
                                        "Key": "code.filepath",
                                        "Value": {
                                                "Type": "STRING",
                                                "Value": "/Users/drew/src/zazen/telemetry/zap_log.go"
                                        }
                                },
                                {
                                        "Key": "code.lineno",
                                        "Value": {
                                                "Type": "INT64",
                                                "Value": 25
                                        }
                                }
                        ],
                        "DroppedAttributeCount": 0,
                        "Time": "2022-05-03T19:28:58.196763-06:00"
                }
        ],
        "Links": null,
        "Status": {
                "Code": "Unset",
                "Description": ""
        },
        "DroppedAttributes": 0,
        "DroppedEvents": 0,
        "DroppedLinks": 0,
        "ChildSpanCount": 0,
        "Resource": [
                {
                        "Key": "service.name",
                        "Value": {
                                "Type": "STRING",
                                "Value": "the-birds"
                        }
                },
                {
                        "Key": "service.version",
                        "Value": {
                                "Type": "STRING",
                                "Value": "0.0.1"
                        }
                }
        ],
        "InstrumentationLibrary": {
                "Name": "github.com/riandyrn/otelchi",
                "Version": "semver:1.0.0",
                "SchemaURL": ""
        }
}
```

The trace/span id's are effectively the uuid you're looking for (though they don't make it to the logs). Any OTLP aware collector will do the right thing with these.